### PR TITLE
Disallow using direction flags for spacedim>dim+1.

### PR DESCRIPTION
**Pull request by bangerth**
This gets me 80% towards #16343. Let's see how much breakage this produces :-)

### CODE REVIEW

File: source/grid/tria_accessor.cc [hunk 1]
```diff
@@ -2225,11 +2225,14 @@ bool
...
-  if (dim == spacedim)
+  if constexpr (dim == spacedim)
     return true;
-  else
+  else if constexpr (dim == spacedim - 1)
     return this->tria->levels[this->present_level]
       ->direction_flags[this->present_index];
+  else
+    Assert(false,
+           ExcMessage("This function cannot be called if dim<spacedim-1."));
```
Comment (drwells) [inline source/grid/tria_accessor.cc:2235]: We need a dummy return here.

File: include/deal.II/grid/tria_accessor.h [hunk 2]
```diff
@@ -3858,7 +3858,10 @@ class CellAccessor : public TriaAccessor<dim, dim, spacedim>
    */
 
   /**
-   * Return the orientation of this cell.
+   * Return the orientation of this cell. This function always returns
+   * `true` if `dim==spacedim`. It can return `true` or `false` if
+   * `dim==spacedim-1`. The function cannot be called (and will throw
+   * an exception) if called for `dim<spacedim-1`.
```
Comment (masterleinad) [inline include/deal.II/grid/tria_accessor.h:3864]: We are calling `Assert` so we would normally abort instead of throwing an exception.

Comment (bangerth) [reply include/deal.II/grid/tria_accessor.h:3864]: We use this same phrase in many many places in the documentation even in places where we use `Assert` internally. But you're right, this shouldn't mean that we should continue this practice. So I fixed this in both places.

**Review by masterleinad — APPROVED** [review]

**Comment by bangerth** [comment]
I fixed a place where we were still calling the now disallowed function. It seems to work now.

**Pull request merged**